lib/Dialect/FIRRTL/Transforms/SpecializeOption.cpp

```diff
@@ -0,0 +1,65 @@
+//===- SpecializeOption.cpp -------------------------------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "PassDetails.h"
+#include "mlir/IR/ImplicitLocOpBuilder.h"
+#include "mlir/IR/Threading.h"
+
+using namespace mlir;
+using namespace circt;
+using namespace firrtl;
+
+namespace {
+struct SpecializeOptionPass
+    : public SpecializeOptionBase<SpecializeOptionPass> {
+  using SpecializeOptionBase::selectedCase;
+  using SpecializeOptionBase::selectedOption;
+
+  void runOnOperation() override {
+    auto circuit = getOperation();
+    if (selectedOption.empty() || selectedCase.empty()) {
+      markAllAnalysesPreserved();
+      return;
+    }
+
+    auto optionGroup = circuit.lookupSymbol<OptionOp>(selectedOption);
+    if (!optionGroup) {
+      markAllAnalysesPreserved();
+      return;
+    }
+
+    auto optionCase = optionGroup.lookupSymbol<OptionCaseOp>(selectedCase);
+    if (!optionCase) {
+      markAllAnalysesPreserved();
+      return;
+    }
```

Comment:
I don't think this is a great API---if these fail, then there's a user error and things arguably shouldn't proceed. Specifically, for something like the user specifies an option/case that doesn't exist or misspells it. If the pass silently does a no-op, then it's difficult to debug as it looks like the pass did the wrong thing.

tl;dr: more `signalPassFailure()` and tests of this behavior. 😉

Reply:
We might set up a flow for all designs and ask for specialization for a target, but a smaller input might have no use of options at all, in which case the pass would fail. I think it is reasonable to turn this into a no-op if there is no option to specialize.